paywall/src/components/Paywall.js

```diff
@@ -7,69 +7,48 @@ import DeveloperOverlay from './developer/DeveloperOverlay'
 import ShowWhenLocked from './lock/ShowWhenLocked'
 import ShowWhenUnlocked from './lock/ShowWhenUnlocked'
 import GlobalErrorProvider from '../utils/GlobalErrorProvider'
-import { lockPage, unlockPage } from '../services/iframeService'
 import { UnlockedFlag } from './lock/UnlockFlag'
 import { lockRoute } from '../utils/routes'
+import useListenForPostMessage from '../hooks/browser/useListenForPostMessage'
+import usePostMessage from '../hooks/browser/usePostMessage'
 
-class Paywall extends React.Component {
-  constructor(props) {
-    super(props)
-    this.state = {
-      scrollPosition: 0,
-    }
-    this.handleScrollPosition = this.handleScrollPosition.bind(this)
-    this.handleIframe = this.handleIframe.bind(this)
-  }
-  componentDidMount() {
-    window.addEventListener('message', event => {
-      if (event.data.scrollPosition) {
-        this.handleScrollPosition(event.data.scrollPosition)
+export function Paywall({ locks, locked, redirect, window }) {
+  const data = useListenForPostMessage(window) || { scrollPosition: 0 }
+  const scrollPosition = data.scrollPosition || 0
```

Comment:
Can lines 16 and 17 be joined? Or is it possible that `useListenForPostMessage` on `window` will return a truthy value that doesn't contain `scrollPosition`? More a question of curiosity, other than that looks good!

Reply:
`useListenForPostMessage` returns whatever the data value is. If it is not an object, the scrollPosition will not be present, and we'll error. Most commonly, on startup it is `null` prior to receiving a message.

The other thing is perhaps that it should be validated, this is a poor man's validation. We actually don't know for sure if the message will come from `paywall.min.js` or from some other script running in the main page, so we have to be careful about any assumptions. Do you have a better idea with this info?

Reply:
I have an idea, not sure if it's reasonable. If we tell `useListenForPostMessage` what data we're expecting, maybe it can always return an object that has default values for the data we want.

Just to spitball:
```
useListenForPostMessage = (window, dataAndDefaults) => {
  // Will always return an object containing the keys of `dataAndDefaults`,
  // if a key is not present in the data value, it takes the default value
  ...
}

// call it like this:
const { scrollPosition } = useListenForPostMessage(window, { scrollPosition: 0 })
```

I don't know if that's the ideal abstraction for this, but it might clean up the code a bit by hoisting the `thing || defaultForThing` into the hook so that the call sites are a little simpler

I don't think this is a critical change right now though, maybe just something to keep in mind if our use of this hook (or others like it) grows to use more properties.